File: src/main/java/com/jackmeng/stl/stl_DisjointSet.java
package com.jackmeng.stl;

public class stl_DisjointSet
{
    private int[] parent, rank;

    public stl_DisjointSet(int size)
    {
        parent = new int[size];
        rank = new int[size];

        for(int i = 0; i < size; i++)
        {
            parent[i] = i;
            rank[i] = 0;
        }
    }

    public int find(int x)
    {
        if(parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    public void union(int x, int y)
    {
        int root_x = find(x);
        int root_y = find(y);
        if(root_x == root_y) return;

        if(rank[root_x] < rank[root_y]) parent[root_x] = root_y;
        else if (rank[root_x] > rank[root_y]) parent[root_y] = root_x;
        else
        {
            parent[root_y] = root_x;
            rank[root_x]++;
        }
    }
}

File: src/main/java/com/jackmeng/stl/stl_DisjointSet.kt
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
package com.jackmeng.stl

class stl_DisjointSet(size:Int)
{
	private val parent:IntArray
	private val rank:IntArray
	
	init
	{
		parent=IntArray(size)
		rank=IntArray(size)
		for (i in 0 until size)
		{
			parent[i]=i
			rank[i]=0
		}
	}
	
	fun find(x:Int):Int
	{
		if (parent[x]!=x) parent[x]=find(parent[x])
		return parent[x]
	}
	
	fun union(x:Int , y:Int)
	{
		val root_x=find(x)
		val root_y=find(y) // ??
		if (root_x==root_y) return
		if (rank[root_x]<rank[root_y]) parent[root_x]=root_y
		else if (rank[root_x]>rank[root_y]) parent[root_y]=root_x
		else
		{
			parent[root_y]=root_x
			rank[root_x]++
		}
	}
}